File: packages/codeact/src/executor.ts
import { createSandboxBackend } from "@mantle/sandbox";
import type { SandboxBackend } from "@mantle/sandbox";

export interface CodeActRequest {
  code: string;
  language?: "python";
}

export interface CodeActResult {
  stdout: string;
  stderr: string;
  results: unknown[];
  files_changed: string[];
  duration_ms: number;
  exit_code: number;
}

interface FileSnapshot {
  [path: string]: string;
}

const SNAPSHOT_CODE = String.raw`
from __future__ import annotations

import hashlib
import json
from pathlib import Path

root = Path("/workspace")
snapshot: dict[str, str] = {}
for path in sorted(root.rglob("*")):
    if not path.is_file() or path.name == ".codeact_tmp.py":
        continue
    digest = hashlib.sha256(path.read_bytes()).hexdigest()
    snapshot[str(path)] = digest
print(json.dumps(snapshot, sort_keys=True))
`;

async function snapshotWorkspace(backend: SandboxBackend, sessionId: string): Promise<FileSnapshot> {
  const result = await backend.exec(sessionId, SNAPSHOT_CODE);
  if (result.exitCode !== 0) {
    throw new Error(`Failed to snapshot sandbox workspace: ${result.stderr}`);
  }
  const parsed: unknown = JSON.parse(result.stdout || "{}");
  if (typeof parsed !== "object" || parsed === null || Array.isArray(parsed)) {
    throw new Error("Sandbox workspace snapshot was not an object");
  }

  const snapshot: FileSnapshot = {};
  for (const [path, digest] of Object.entries(parsed)) {
    if (typeof digest !== "string") {
      throw new Error(`Sandbox workspace snapshot digest for ${path} was not a string`);
    }
    snapshot[path] = digest;
  }
  return snapshot;
}

function diffSnapshots(before: FileSnapshot, after: FileSnapshot): string[] {
  const paths = new Set([...Object.keys(before), ...Object.keys(after)]);
  return [...paths].filter((path) => before[path] !== after[path]).sort();
}

function parseResults(stdout: string): unknown[] {
  const marker = "__MANTLE_RESULT__=";
  return stdout
    .split("\n")
    .filter((line) => line.startsWith(marker))
    .map((line) => JSON.parse(line.slice(marker.length)) as unknown);
}

export async function execute_code(request: CodeActRequest): Promise<CodeActResult> {
  if (request.language && request.language !== "python") {
    throw new Error(`Unsupported CodeAct language: ${request.language}`);
  }

  const backend = createSandboxBackend();
  const startedAt = Date.now();
  let sessionId: string | null = null;

  try {
    sessionId = await backend.create();
    const before = await snapshotWorkspace(backend, sessionId);
    const execResult = await backend.exec(sessionId, request.code);
    const after = await snapshotWorkspace(backend, sessionId);

    return {
      stdout: execResult.stdout,
      stderr: execResult.stderr,
      results: parseResults(execResult.stdout),
      files_changed: diffSnapshots(before, after),
      duration_ms: Date.now() - startedAt,
      exit_code: execResult.exitCode,
    };
  } finally {
    if (sessionId) {
      await backend.destroy(sessionId);
    }
  }
}

File: packages/codeact/codeact/executor.py
from __future__ import annotations

import json
import time
from dataclasses import dataclass
from typing import Any

from sandbox.backend import SandboxBackend, create_backend


@dataclass
class CodeActResult:
    stdout: str
    stderr: str
    results: list[Any]
    files_changed: list[str]
    duration_ms: int
    exit_code: int


SNAPSHOT_CODE = r'''
from __future__ import annotations

import hashlib
import json
from pathlib import Path

root = Path("/workspace")
snapshot: dict[str, str] = {}
for path in sorted(root.rglob("*")):
    if not path.is_file() or path.name == ".codeact_tmp.py":
        continue
    snapshot[str(path)] = hashlib.sha256(path.read_bytes()).hexdigest()
print(json.dumps(snapshot, sort_keys=True))
'''


def _snapshot_workspace(backend: SandboxBackend, session_id: str) -> dict[str, str]:
    result = backend.exec_code(session_id, SNAPSHOT_CODE)
    if result.exit_code != 0:
        raise RuntimeError(f"Failed to snapshot sandbox workspace: {result.stderr}")
    loaded = json.loads(result.stdout or "{}")
    if not isinstance(loaded, dict):
        raise RuntimeError("Sandbox workspace snapshot was not an object")
    return {str(path): str(digest) for path, digest in loaded.items()}


def _diff_snapshots(before: dict[str, str], after: dict[str, str]) -> list[str]:
    return sorted(path for path in set(before) | set(after) if before.get(path) != after.get(path))


def _parse_results(stdout: str) -> list[Any]:
    marker = "__MANTLE_RESULT__="
    results: list[Any] = []
    for line in stdout.splitlines():
        if line.startswith(marker):
            results.append(json.loads(line[len(marker) :]))
    return results


async def execute_code(code: str, language: str = "python") -> CodeActResult:
    if language != "python":
        raise ValueError(f"Unsupported CodeAct language: {language}")

    backend = create_backend()
    started_at = time.monotonic()
    session_id: str | None = None

    try:
        session_id = backend.create()
        before = _snapshot_workspace(backend, session_id)
        exec_result = backend.exec_code(session_id, code)
        after = _snapshot_workspace(backend, session_id)
        return CodeActResult(
            stdout=exec_result.stdout,
            stderr=exec_result.stderr,
            results=_parse_results(exec_result.stdout),
            files_changed=_diff_snapshots(before, after),
            duration_ms=int((time.monotonic() - started_at) * 1000),
            exit_code=exec_result.exit_code,
        )
    finally:
        if session_id is not None:
            backend.destroy(session_id)
